include/GcInfo/GcInfo.h

```diff
@@ -29,46 +29,103 @@
+// GC Values as untracked.
+// These stack allocations are known to be live throughout the function,
+// because the reader marks them as frame-escaped.
+
+struct AllocaInfo {
+  int32_t Offset;
+  AllocaFlags Flags; // Combination of AllocaFlags, to be precise.
+  const char *getAllocTypeString() const;
+  bool isGcPointer() const { return ((Flags & Alloca_GcPointer) != 0); }
+  bool isGcAggregate() const { return ((Flags & Alloca_GcAggregate) != 0); }
+  bool isGcValue() const {
+    return ((Flags & (Alloca_GcPointer | Alloca_GcAggregate)) != 0);
+  }
+  bool isPinned() const { return ((Flags & Alloca_Pinned) != 0); }
+};
+
+/// \brief Per Function GcInfo
+/// Allocation map for stack allocated GC Values
+/// and special symbols.
 
 class GcFuncInfo {
 public:
   GcFuncInfo(const llvm::Function *F);
 
-  void recordPinnedSlot(llvm::AllocaInst *Alloca);
-  void recordGcAggregate(llvm::AllocaInst *Alloca);
+  void recordGcAlloca(const llvm::AllocaInst *Alloca);
+  void recordPinned(const llvm::AllocaInst *Alloca);
+  void recordSecurityObject(const llvm::AllocaInst *Alloca);
+  void recordGsCookie(const llvm::AllocaInst *Alloca,
+                      const uint32_t ValidRangeStart,
+                      const uint32_t ValidRangeEnd);
+  void recordGenericsContext(const llvm::AllocaInst *Alloca,
+                             const GENERIC_CONTEXTPARAM_TYPE ParamType);
+
+  bool hasRecord(const llvm::AllocaInst *Alloca) {
+    return (AllocaMap.find(Alloca) != AllocaMap.end());
+  }
+
   void getEscapingLocations(llvm::SmallVector<llvm::Value *, 4> &EscapingLocs);
 
   // Function for which GcInfo is recorded
   const llvm::Function *Function;
 
-  // Allocation to frame-offset Mappings
-  llvm::ValueMap<const llvm::AllocaInst *, int32_t> PinnedSlots;
-  llvm::ValueMap<const llvm::AllocaInst *, int32_t> GcAggregates;
-  llvm::AllocaInst *GsCookie;
-  int32_t GsCookieOffset;
+  // Alloca Instruction to AllocaInfo Map, for:
+  // a) All stack allocated GC Values
+  // b) Certain special allocations like Generics Context Parameter,
+  //    which are not of GC-type.
+  llvm::ValueMap<const llvm::AllocaInst *, AllocaInfo> AllocaMap;
+
+  // Additional information for special slots
   uint32_t GsCkValidRangeStart;
   uint32_t GsCkValidRangeEnd;
-  llvm::AllocaInst *SecurityObject;
-  int32_t SecurityObjectOffset;
-  llvm::AllocaInst *GenericsContext;
-  int32_t GenericsContextOffset;
   GENERIC_CONTEXTPARAM_TYPE GenericsContextParamType;
+
+private:
+  // Record a Stack Allocation in the FuncInfo, with appropriate
+  // Flags based on Type of allocation.
+  void recordAlloca(const llvm::AllocaInst *Alloca);
+
+  // Mark additional annotations on a recorded GC Value
+  void markGcAlloca(const llvm::AllocaInst *Alloca, const AllocaFlags Flags);
+
+  // Mark additional annotations on a recorded non-GC-Value
+  void markIntAlloca(const llvm::AllocaInst *Alloca, const AllocaFlags Flags);
```

Comment:
I just meant it seems like there are two cases, recording a `gc` and recording something that is not `gc` (we don't really care what it is, just that it does not need gc reporting), so the naming might reflect that more directly.

For the `const` case: if you want to ensure that the parameter value is not modified by the method, you can put the `const` only on the definition and leave it off the declaration. Yes, it evidently works. It might seem odd but more directly captures your intent.